doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,653 @@
+other than `SSL_want_net_read`, `SSL_want_net_write`,
+`SSL_get_rpoll_descriptor`, `SSL_get_wpoll_descriptor` and
+`SSL_get_tick_timeout`.
+
+#### `SSL_want`, `SSL_want_read`, `SSL_want_write`
+
+The existing API `SSL_want`, and the macros defined in terms of it, are
+traditionally used to determine if the SSL state machine has exited in
+non-blocking mode due to a desire to read from or write to the underlying
+network BIO. However, this API is unsuitable for use with QUIC because the
+return value of `SSL_want` can only express one I/O direction at a time (read or
+write), not both. This call will not be implemented for QUIC (e.g. always
+returns `SSL_NOTHING`) and `SSL_want_net_read` and `SSL_want_net_write` will be
+used instead.
+
+TBD: Should these be implemented for non-QUIC SSL objects?
+
+#### `SSL_set_initial_peer_addr`, `SSL_get_initial_peer_addr`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| New       | Never         | No        | CS            |
+
+`SSL_set_initial_peer_addr` sets the initial L4 UDP peer address for an outgoing
+QUIC connection.
+
+The initial peer address may be autodetected if no peer address has already been
+set explicitly and the QUIC connection SSL object is provided with a
+`BIO_s_dgram` with a peer set.
+
+`SSL_set_initial_peer_addr` cannot be called after a connection is established.
+
+### Future APIs
+
+A custom poller interface may be provided in the future. For more information,
+see the QUIC I/O Architecture design document.
+
+BIO Objects
+-----------
+
+### Existing APIs
+
+#### `BIO_s_connect`, `BIO_new_ssl_connect`, `BIO_set_conn_hostname`
+
+We are aiming to support use of the existing `BIO_new_ssl_connect` API with only
+minimal changes. This will require internal changes to `BIO_s_connect`, which
+should automatically detect when it is being used with a QUIC `SSL_CTX` and act
+accordingly.
+
+#### `BIO_new_bio_pair`
+
+Unsuitable for use with QUIC on the network side; instead, applications can
+make use of the new `BIO_s_dgram_pair` which provides equivalent functionality
+with datagram semantics.
+
+#### Interactions with `BIO_f_buffer`
+
+Existing applications sometimes combine a network socket BIO with a
+`BIO_f_buffer`. This is problematic because the datagram semantics of writes are
+not preserved, therefore the BIO provided to libssl is, as provided, unusable
+for the purposes of implementing QUIC. Moreover, output buffering is not a
+relevant or desirable performance optimisation for the transmission of UDP
+datagrams and will actually undermine QUIC performance by causing incorrect
+calculation of ACK delays and consequently inaccurate RTT calculation.
+
+Options:
+
+  - Require applications to be changed to not use QUIC with a `BIO_f_buffer`.
+  - Detect when a `BIO_f_buffer` is part of a BIO stack and bypass it
+    (yucky and surprising).
```

Comment:
We already have `BIO_TYPE_*` values, some of which are used to define "classes" of BIOs:

````
# define BIO_TYPE_DESCRIPTOR     0x0100 /* socket, fd, connect or accept */
# define BIO_TYPE_FILTER         0x0200
# define BIO_TYPE_SOURCE_SINK    0x0400
````

You could envisage a "stream" class and a "datagram" class. Joining a "stream" BIO to a "datagram" BIO via `BIO_push` could be an error. Although that might cause backwards compat problems.

Reply:
Right, applications are going to have to *not* do this for QUIC, basically. It is what it is. Ideally we should try and detect when an application is trying to do this and fail noisily though.

Reply:
The filter BIO has to support the BIO_sendmmsg BIO_recvmmsg calls - that can be easily detected, can't it?